src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+            H5T__bit_set(d + dst_part_size, dst_atomic.u.f.epos, dst_atomic.u.f.esize, false);
+            H5T__bit_set(d, dst_atomic.u.f.mpos, dst_atomic.u.f.msize, false);
+            H5T__bit_set(d + dst_part_size, dst_atomic.u.f.mpos, dst_atomic.u.f.msize, false);
+            goto padding;
+        }
+        else if (real_except || imag_except) {
+            except_ret = H5T_CONV_UNHANDLED;
+
+            /* If user's exception handler is present, use it */
+            if (conv_ctx->u.conv.cb_struct.func) {
+                H5T_conv_except_t except_type; /* type of conversion exception that occurred */
+
+                /* reverse source buffer order first */
+                H5T__reverse_order(src_rev, s, src_p);
+
+                /*
+                 * A complex number is infinity if either part is infinity,
+                 * even if the other part is NaN. If a part is infinity,
+                 * since we can only throw one type of conversion exception,
+                 * arbitrarily choose the exception type to throw based
+                 * on the infinity type for the real part (if it's infinity),
+                 * followed by the infinity type for the imaginary part. For
+                 * now, it will be assumed that the conversion exception
+                 * callback will inspect and handle both parts of the complex
+                 * number value.
+                 */
+                if (realval_type == H5T_CONV_FLOAT_SPECVAL_POSINF)
+                    except_type = H5T_CONV_EXCEPT_PINF;
+                else if (realval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF)
+                    except_type = H5T_CONV_EXCEPT_NINF;
+                else if (imagval_type == H5T_CONV_FLOAT_SPECVAL_POSINF)
+                    except_type = H5T_CONV_EXCEPT_PINF;
+                else if (imagval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF)
+                    except_type = H5T_CONV_EXCEPT_NINF;
+                else {
+                    assert(realval_type == H5T_CONV_FLOAT_SPECVAL_NAN ||
+                           imagval_type == H5T_CONV_FLOAT_SPECVAL_NAN);
+                    except_type = H5T_CONV_EXCEPT_NAN;
+                }
+
+                except_ret = (conv_ctx->u.conv.cb_struct.func)(except_type, conv_ctx->u.conv.src_type_id,
+                                                               conv_ctx->u.conv.dst_type_id, src_rev, d,
+                                                               conv_ctx->u.conv.cb_struct.user_data);
+            }
+
+            if (except_ret == H5T_CONV_UNHANDLED) {
+                if (realval_type == H5T_CONV_FLOAT_SPECVAL_POSINF ||
+                    realval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF ||
+                    imagval_type == H5T_CONV_FLOAT_SPECVAL_POSINF ||
+                    imagval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF) {
+                    H5T__bit_copy(d, dst_atomic.u.f.sign, s, src_atomic.u.f.sign, (size_t)1);
```

Comment:
Could you copy the global infinity values here?

Reply:
Sure, as long as the type being converted matches one of the native types. Would probably make it a bit uglier here with special-case code and realistically this function is only ever going to be used for non-standard complex types like a complex of float16 anyway. I'm sure several of the library's software conversion functions could be heavily optimized like this, though I don't know if it's worth doing for this PR.